Subscriber groups/shared subscriptions

larshesel opened the pull request:
@dergraf Ready for review

 - policies implemented: `random` and `prefer_local`. We could consider adding `only_local` as well.
 - enqueue_many with options only implemented in 'online' state. Should probably be implemented in others as well?
 - needs to be verified against the MQTTv5 spec.
 - must be squashed before merging.

dergraf commented on apps/vmq_server/src/vmq_cluster_com.erl:
```diff
@@ -142,34 +142,58 @@ process_bytes(Bytes, Buffer, St) ->
+                          try
+                              case vmq_queue_sup_sup:get_queue_pid(SubscriberId) of
+                                  QueuePid when is_pid(QueuePid) ->
+                                      %% TODO, maybe replace with dedicated
```
what is this?

larshesel replied:
removed.

dergraf commented on apps/vmq_server/src/vmq_reg.erl:
```diff
@@ -268,6 +268,61 @@ register_session(SubscriberId, QueueOpts) ->
+    publish_to_subscriber_groups(Msg, Policy, maps:to_list(SubscriberGroups));
+publish_to_subscriber_groups(_,_, []) -> ok;
+publish_to_subscriber_groups(Msg, Policy, [{Group, []}|Rest]) ->
+    lager:info("can't publish to subscriber group ~p due to no subscriber available", [Group]),
```
debug instead of info? If debug add the Msg to the statement.

larshesel replied:
fixed and reworded.

dergraf commented on apps/vmq_server/src/vmq_reg.erl:
```diff
@@ -268,6 +268,61 @@ register_session(SubscriberId, QueueOpts) ->
+                ok ->
+                    publish_to_subscriber_groups(Msg, Policy, Rest);
+                {error, Reason} ->
+                    lager:warning("can't publish to subscriber group on remote node ~p due to '~p'",
```
Maybe debug instead of warning

larshesel replied:
fixed and reworded.

dergraf commented:
LGTM! :)

larshesel commented:
Squashed and rebased the subscriber group onto master and changed the message compatibility layer so the cluster com is compatible with pre-subscriber group nodes (need to test this)

larshesel commented:
Link this PR to the related issues: #230 and #221

dergraf commented:
I think as the official spec isn't final yet, this is good to merge, we can change it later, and mark this feature as beta. 

comments?

larshesel commented:
I think that's fair. I'll do some mixed-cluster testing and merge if that looks OK. Then I'll start documenting it (and include a note about the beta status etc).

larshesel commented:
Hi @dergraf it looks good to me - my mixed-cluster tests didnt't reveal any issues with the cluster comm. I added an intermediate entry to the changelog so we can document features as we add them and then we'd only need to change the heading when we do the actual release.

Could you take a quick look at the changelog? If the content there is OK with you I'll merge it (or feel free to do it)

ig0r commented:
guys,

Is it included in the release? 

Trying it in 0.15.3 and don't see it's working

ig0r commented:
Oh, I see. Thanks for the heads-up!
